Composable filters and deeply nested queries

bart-degreed opened the pull request:
_"Everybody knew it was impossible, until someone came along that didn't know."_

# Overview

I'm proud to present this PR, which brings the following enhancements:

**1. Composable filter expressions**
Examples:
- /api/v1/users?filter=and(equals(lastName,'O''Brian'),greaterOrEqual(age,'25'),lessThan(age,'35'))
	Returns all users with last name O'Brian and age in range 25-35.
- /api/v1/users?filter=or(endsWith(lastName,'Conner'),any(lastName,'Smith','Williams','Miller'))
	Returns users whose last name either ends in Conner or equals Smith, Williams or Miller.
- /api/v1/blogs?filter=has(articles)&include=articles&filter[articles]=not(equals(publishDate,null))
	Returns blogs that have at least one article. Only includes articles that have a publication date.
- /api/v1/blogs/1/articles?filter=greaterThan(count(upvotes),count(downvotes))
	Returns articles from blog 1 where the number of upvotes exceeds the number of downvotes.

**2. Deeply nested filtering, pagination, sorting, and sparse fieldsets**
Examples:
- /api/v1/blogs?include=articles.comments&page[number]=2,articles:3,articles.comments:4&page[size]=10,articles:5
	Returns at most 10 blogs at page 2. Per blog, returns at most 5 included articles, all at page 3. Per blog, per article, returns included comments at page 4 using default page size.
- /api/v1/blogs?sort=title,-id&include=articles&sort[articles]=-publishDate
	Sorts blogs by title, then ID in descending order. Sorts included articles descending by publication date.
- /api/v1/blogs/1/articles?fields=summary&include=comments&fields[comments]=title,userName
	Returns article summaries from blog with ID 1. Returns titles and usernames from included comments.

**3. New extensibility points in ResourceDefinitions to adapt, replace or discard query string input**
- `virtual IReadOnlyCollection<IncludeElementExpression> OnApplyIncludes(IReadOnlyCollection<IncludeElementExpression> existingIncludes)`
- `virtual FilterExpression OnApplyFilter(FilterExpression existingFilter)`
- `virtual PaginationExpression OnApplyPagination(PaginationExpression existingPagination)`
- `virtual SortExpression OnApplySort(SortExpression existingSort)`
- `virtual SparseFieldSetExpression OnApplySparseFieldSet(SparseFieldSetExpression existingSparseFieldSet)`

But the best part is:
### All of these get translated into SQL queries, nothing is evaluated in memory!

**4. Service/repository separation**
- `IQueryable<T>` is only used internally in repository, enabling service reuse for alternate data stores.

**5. Additional minor enhancements**
- Use `~AttrCapabilities.AllowView` to omit attribute from responses; if turned off, fails with HTTP 400 when requested in `?fields=`
- When deleting a resource, it no longer fetches the entity first
- `ICurrentRequest.IsReadOnly` enables to use a different connection string for read-only requests
- Optimized field selection on nested endpoints and relationships: for the parent, only retrieve `Id` column
- Improved integration tests: better isolation using temporary database per test class (`IntegrationTestContext`) and cascaded table truncate (`ClearTableAsync`)

# So how does it work?
The query pipeline roughly looks like this:

```
HTTP --[ASP.NET Core]--> QueryString --[JADNC:QueryStringParameterReader]--> QueryExpression[] --[JADNC:ResourceService]--> QueryLayer --[JADNC:Repository]--> IQueryable --[EF Core]--> SQL
```

Processing a request involves the following steps:
- `JsonApiMiddleware` collects resource info from routing data for the current request. **(unchanged)**
- `JsonApiReader` transforms json request body into objects. **(unchanged)**
- `JsonApiController` accepts get/post/patch/delete verb and delegates to service. **(unchanged)**
- `IQueryStringParameterReader`s delegate to `QueryParser`s that transform query string text into `QueryExpression` objects. **(new)**
	- By using prefix notation in filters, we don't need users to remember operator precedence and associativity rules.
	- These validated expressions contain direct references to attributes and relationships.
	- The readers also implement `IQueryConstraintProvider`, which exposes expressions through `ExpressionInScope` objects.
- `QueryLayerComposer` (used from `JsonApiResourceService`) collects all query constraints. **(new)**
	- It combines them with default options and `ResourceDefinition` overrides and composes a tree of `QueryLayer` objects.
	- It lifts the tree for nested endpoints like /blogs/1/articles and rewrites includes.
	- `JsonApiResourceService` contains no more usage of `IQueryable`.
- `EntityFrameworkCoreRepository` delegates to `QueryableBuilder` to transform the `QueryLayer` tree into `IQueryable` expression trees. **(new)**
	`QueryBuilder` depends on `QueryClauseBuilder` implementations that visit the tree nodes, transforming them to `System.Linq.Expression` equivalents.
	The `IQueryable` expression trees are executed by EF Core, which produces SQL statements out of them.

# Example
To get a sense of what this all looks like, let's look at an example query string:

```
/api/v1/blogs?
  include=owner,articles.revisions.author&
  filter=has(articles)&
  sort=count(articles)&
  page[number]=3&
  fields=title&
    filter[articles]=and(not(equals(author.firstName,null)),has(revisions))&
    sort[articles]=author.lastName&
    fields[articles]=url&
      filter[articles.revisions]=and(greaterThan(publishTime,'2001-01-01'),startsWith(author.firstName,'J'))&
      sort[articles.revisions]=-publishTime,author.lastName&
      fields[articles.revisions]=publishTime
```

After parsing, the set of scoped expressions is transformed into the following tree by `QueryLayerComposer`:

```
QueryLayer<Blog>
{
  Include: owner,articles.revisions
  Filter: has(articles)
  Sort: count(articles)
  Pagination: Page number: 3, size: 5
  Projection
  {
    title
    id
    owner: QueryLayer<Author>
    {
      Sort: id
      Pagination: Page number: 1, size: 5
    }
    articles: QueryLayer<Article>
    {
      Filter: and(not(equals(author.firstName,null)),has(revisions))
      Sort: author.lastName
      Pagination: Page number: 1, size: 5
      Projection
      {
        url
        id
        revisions: QueryLayer<Revision>
        {
          Filter: and(greaterThan(publishTime,'2001-01-01'),startsWith(author.firstName,'J'))
          Sort: -publishTime,author.lastName
          Pagination: Page number: 1, size: 5
          Projection
          {
            publishTime
            id
          }
        }
      }
    }
  }
}
```

Next, the repository translates this into a LINQ query that the following C# code would represent:

```c#
var query = dbContext.Blogs
    .Include("Owner")
    .Include("Articles.Revisions")
    .Where(blog => blog.Articles.Any())
    .OrderBy(blog => blog.Articles.Count)
    .Skip(10)
    .Take(5)
    .Select(blog => new Blog
    {
        Title = blog.Title,
        Id = blog.Id,
        Owner = blog.Owner,
        Articles = new List<Article>(blog.Articles
            .Where(article => article.Author.FirstName != null && article.Revisions.Any())
            .OrderBy(article => article.Author.LastName)
            .Take(5)
            .Select(article => new Article
            {
                Url = article.Url,
                Id = article.Id,
                Revisions = new HashSet<Revision>(article.Revisions
                    .Where(revision => revision.PublishTime > DateTime.Parse("2001-01-01") && revision.Author.FirstName.StartsWith("J"))
                    .OrderByDescending(revision => revision.PublishTime)
                    .ThenBy(revision => revision.Author.LastName)
                    .Take(5)
                    .Select(revision => new Revision
                    {
                        PublishTime = revision.PublishTime,
                        Id = revision.Id
                    }))
            }))
    });
```

# Breaking changes

- New filter query string syntax. Set `options.EnableLegacyFilterNotation` to `true` to allow legacy filters
	To use new notation, prefix with "expr:", for example: `?filter=expr:equals(lastName,'Smith')`
- Multiple filters in query string _at same depth_ are combined using OR operator (used to be AND, which violates json:api recommendations)
- Using a negative page number (to reverse order) is no longer possible
- "total-records" in response meta has been renamed to "total-resources" (and casing convention is applied)
- `ResourceDefinition<T>.HideFields()` has been replaced by `ResourceDefinition<T>.OnApplySparseFieldSet()`
- `ResourceDefinition<T>.GetQueryFilters()` has been replaced by `ResourceDefinition<T>.OnRegisterQueryableHandlersForQueryStringParameters`
	These are no longer tied to only filters. For example: `?filter[isHighRisk]=true` now uses: `?isHighRisk=true`
- When no sort is provided, resources are sorted ascending by ID
- Notable renames:
	`DefaultResourceService` -> `JsonApiResourceService`
	`DefaultResourceRepository` -> `EntityFrameworkCoreRepository`
	`BaseJsonApiController.GetRelationshipsAsync` -> `GetRelationshipAsync`
	`BaseJsonApiController.GetRelationshipAsync` -> `GetSecondaryAsync`
	`AttrCapabilities.AllowMutate` -> `AllowChange`
- Most occurrences of 'entity' were renamed to 'resource' and `Default` prefix was removed from various class names
- Unified internal terminology: main=base (/blogs) -> primary, nested=relationship=deep (/blogs/1/owner) -> secondary

# Legacy filter conversion table

| Old					| New						|
|---------------------------------------|-----------------------------------------------|
| `?filter[attribute]=value`		| `?filter=equals(attribute,'value')`		|
| `?filter[attribute]=ne:value`		| `?filter=not(equals(attribute,'value'))`	|
| `?filter[attribute]=lt:10`		| `?filter=lessThan(attribute,'10')`		|
| `?filter[attribute]=gt:10`		| `?filter=greaterThan(attribute,'10')`		|
| `?filter[attribute]=le:10`		| `?filter=lessOrEqual(attribute,'10')`		|
| `?filter[attribute]=ge:10`		| `?filter=greaterOrEqual(attribute,'10')`	|
| `?filter[attribute]=like:value`		| `?filter=contains(attribute,'value')`		|
| `?filter[attribute]=in:value1,value2`	| `?filter=any(attribute,'value1,'value2')`	|
| `?filter[attribute]=nin:value1,value2`	| `?filter=not(any(attribute,'value1,'value2'))`	|
| `?filter[attribute]=isnull:`		| `?filter=equals(attribute,null)`		|
| `?filter[attribute]=isnotnull:`		| `?filter=not(equals(attribute,null))`		|

# Pending work

- [x] address some refacoring TODOs in code
- [x] code review: namespaces and type names, type/member accessibility and extensibility
- [ ] explore customizations
	- use case: custom filter operator: 'range' (inherit tokenizer/parser/builder etc)
	- use case: replace EF Core Repository with different data store
- [ ] assess performance changes, consider different collection types, caching for parsed queries (they are immutable)

# Fixed issues

Fixes #788
Fixes #787
Fixes #764
Fixes #761
Fixes #758
Fixes #757
Fixes #751
Fixes #748
Fixes #747
Fixes #738
Fixes #551
Fixes #444
Fixes #217
Fixes #183
Fixes #176

This PR supersedes the following PRs: #705, #782.

# Known limitations

These bugs in EF Core prevent some scenarios from working correctly. Please upvote them if they are important for you.
* https://github.com/dotnet/efcore/issues/21026
* https://github.com/dotnet/efcore/issues/20502

bart-degreed commented:
The coming weeks I won't be available to respond, but please feel free to try things out and provide feedback.

ClintGood commented:
Hi Bart

All this looks very clean and powerful.

The following may not be directly related to this PR but maybe a bigger issue. I can open up another issue on the main fork if that makes sense. Anyway...

I tested against my model which does not have a default "Id" field in the underlying table.
My model implements IIdentifiable<T>
There is code that implicitly assumes there will be an Id field in the class and will fail if it is missing (eg. QueryLayerComposer.GetSort)

Attempting to work around this by creating a non mapped Id field will get past the above problem, but will incorrectly try and filter on an Id field in the database.

In short, there seems to be code that always expects Identifiable instead of IIdentifiable<T>

bart-degreed commented:
Hi @ClintGood thanks for your feedback. As far as I know, there has always been an implicit requirement for a typed Id property to exist. This is something the json:api specification depends on. I've added a comment at #797 that may solve your problem.

bjornharrtell commented:
Impressive work @bart-degreed. :+1:

maurei commented:
General feedback item: all `JsonApiDotNetCore.Internal.Queries.Expressions` could use some explanatory docs 👍  A lot of new concepts are being introduced there, will be of great help for any dev to understand/maintain the internals

bart-degreed commented:
@maurei I've moved several typed out of Internal namespace and added doc-comments. Please let me know specific locations where you believe more explanation would help. I'll address overview documentation later.

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Expressions/LogicalOperator.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace JsonApiDotNetCore.Internal.Queries.Expressions
+{
+    public enum LogicalOperator
```
The members None and All are used on Flags enums (a bitwise combination of multiple members), which this is not. One value must be selected, which is either And or Or. None has no meaning here. The default value of 0 maps to And here, so it is compliant with CA1008:

> An enumeration without an applied System.FlagsAttribute does not define a member that has a value of zero. Or, an enumeration that has an applied FlagsAttribute defines a member that has a value of zero but its name is not 'None'. Or, the enumeration defines multiple, zero-valued members.

maurei commented on test/UnitTests/QueryStringParameters/FilterParseTests.cs:
```diff
@@ -0,0 +1,141 @@
+using System;
```
Type in file name

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/QueryTokenizer.cs:
```diff
@@ -0,0 +1,139 @@
+            return SingleCharacterToTokenKinds.ContainsKey(ch) ? (TokenKind?)SingleCharacterToTokenKinds[ch] : null;
+        }
+
+        private Token FlushTextBuffer(bool isQuotedText)
```
That's basically how any tokenizer works: It loops through the source text and fills up a buffer that holds the data for the next token being built. Then at some point in time, it needs to produce a token from the buffer, before moving onto the next token.

FlushTextBuffer is private and only called from the main tokenizer loop. The main loop could be split into smaller methods to improve readability, but I think it is fine for now. The only complexity in there is dealing with text between quotes (where characters are taken as-is instead of considering them separate tokens).

I don't think this needs any explanation in the code (as I would just be describing what the code does, which is an anti-pattern), but if you feel differently please indicate how you want it changed, as in "add the following text at line....." instead of only asking for docs. I cannot guess what clarification you expect.

bart-degreed replied:
I can rename this method to ProduceTokenFromTextBuffer if that helps?

maurei replied:
The latter is more descriptive and helps me understand the usage of the method 👍

bart-degreed replied:
Renamed.

bart-degreed commented on test/JsonApiDotNetCoreExampleTests/IntegrationTestContext.cs:
```diff
@@ -0,0 +1,201 @@
+            _afterServicesConfiguration = servicesConfiguration;
+        }
+
+        public async Task RunOnDatabaseAsync(Func<TDbContext, Task> asyncAction)
```
This is not something we ship, just used internally by the project itself. When that changes, we should consider adding clarification if needed, because users cannot see the source directly. For now, it contains three lines of code and I think the method name describes exactly what it does, so I see no need for extra explanation. If you disagree, please indicate what extra clarification you need exactly.

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/ResolveFieldChainCallback.cs:
```diff
@@ -0,0 +1,18 @@
+namespace JsonApiDotNetCore.Internal.Queries.Parsing
+{
+    [Flags]
+    public enum FieldChainRequirements
```
I've intentionally left that out, because I'm not happy how it works and may change in the future. It's in the Internal namespace. The only reason it is a public type instead of internal is because it must be, as it is used in a protected method of the parser base class.

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/ResolveFieldChainCallback.cs:
```diff
@@ -0,0 +1,18 @@
+namespace JsonApiDotNetCore.Internal.Queries.Parsing
+{
+    [Flags]
+    public enum FieldChainRequirements
```
Like mentioned earlier, I'm not so happy about this enum at all. And like to shove it under the carpet a bit :)
That's also why I pushed things inside the parsers instead of expecting callers to deal with this. You are right that this violates the rule, but None would have no meaning. This should have been a non-flags enum, but there is one case where multiple values are needed. This is a candidate for refactoring, I'm just not sure what would be the best approach right now.

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/ResolveFieldChainCallback.cs:
```diff
@@ -0,0 +1,18 @@
+        IsRelationship = EndsInToOne | EndsInToMany
+    }
+
+    public delegate IReadOnlyCollection<ResourceFieldAttribute> ResolveFieldChainCallback(string path, FieldChainRequirements requirements);
```
This type no longer exists.

bart-degreed commented on test/JsonApiDotNetCoreExampleTests/IntegrationTests/Filtering/FilterTests.cs:
```diff
@@ -0,0 +1,197 @@
+        }
+
+        [Fact]
+        public async Task Cannot_filter_in_unknown_scope()
```
Your link is just one way for naming tests. The project currently has a mix of styles and I agree we would benefit on standardization. However I think UnitOfWork_StateUnderTest_ExpectedBehavior works best for unit tests, which are usually very technical in nature. As in "when this executes with that parameter, it must return null"

But these are integration tests, which means they spin up a web server and execute a request in order to represent a functional scenario. As in "when filtering recursively it must remove duplicates".  And I think the [given/when/then](https://martinfowler.com/bliki/GivenWhenThen.html) notation better matches that. The shorter form (when/then) I've been using is another common notation, see https://github.com/fluentassertions/fluentassertions/blob/develop/Tests/FluentAssertions.Specs/Numeric/NumericAssertionSpecs.cs for examples.

maurei replied:
Discussed & resolved in private meeting

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/FilterParser.cs:
```diff
@@ -0,0 +1,321 @@
+
+            var expression = ParseFilter();
+
+            AssertTokenStackIsEmpty();
```
Invalid client input. That's why it throws a nicely-formatted QueryParseException instead of some internal error. Getting there is covered by unit tests. It is also a good fallback to have in general whenever a client comes up with a new kind of invalid input that results in the parser thinking it has completed while there was actually more text. simple example: has(articles)XXX

On a side note, I often wonder what some code does. To find out, I comment out something or add `throw null`, run all tests and see which ones fail. That would work here too. Again, please don't just ask for "some comments" but indicate the explanation you believe is missing.

maurei replied:
When I suggest to elaborate on the control flow of a piece of code with a line of comment, it means it is not clear what the control flow is to me and that I do not have a suggestion for that comment, and hence I will keep on asking for "some comments".

In this case, an undocumented method is called from within an undocumented method named `Parse`, where semantically *asserting if empty* at first glance does not have clear overlap with *parsing*, hence some confusion is possible here. Also, the responsibility of catching and handling client side errors is not mentioned in the documentation of `FilterParser` or its base class and is only explicitly clear for me now because of your comment. I'm asking for elaborative comments because I should not have alter the methods body to see which unit tests fail, to deduce these implicit responsibilities of a method/class

bart-degreed replied:
I really think you're pushing this too far. I don't intend to explain what parsing means (reading and validating an input text and producing objects from that while rejecting invalid input) and neither how such a process can be achieved. If unfamiliar with the concept, anyone can look that up on Wikipedia if needed. Loading a stack, then pushing/popping from that while recursing is the usual practice. Take a look at the sources of EF Core or roslyn and you'll see similar, undocumented internals.

My goal is not to explain how internals work in text, which is an anti-pattern because it is only a matter of time to become contradicting with the code. The way to discover how something is implemented (instead of how to use it) is debugging, looking at tests and analyzing usages, like I mentioned. That is in my experience common practice for a developer that wants to understand how *internals* work, in the process of adapting them. So instead I put effort in giving meaningful names over writing text that explains and I believe that's the better option. By not stating how something works internally we preserve the freedom to change that over time without being breaking.

Expressing that you don't understand something is fine, and asking for clarification too. Then once you understand, feel free to propose which comment would be helpful to put in the code and where, so I can either take that, suggest something else or object. Again, asking for "some comments" is not actionable for me, as I cannot image what bits you need clarified and how.

bart-degreed replied:
Update: I've added the following on the parser base class, hope that meets your needs.

```c#
    /// <summary>
    /// The base class for parsing query string parameters, using the Recursive Descent algorithm.
    /// </summary>
    /// <remarks>
    /// Uses a tokenizer to populate a stack of tokens, which is then manipulated from the various parsing routines for subexpressions.
    /// Implementations should throw <see cref="QueryParseException"/> on invalid input.
    /// </remarks>
```

bart-degreed commented on test/UnitTests/QueryStringParameters/FilterParseTests.cs:
```diff
@@ -0,0 +1,141 @@
+using System;
```
I replied to that elsewhere.

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/QueryParser.cs:
```diff
@@ -0,0 +1,79 @@
+
+        protected void EatSingleCharacterToken(TokenKind kind)
+        {
+            if (!TokenStack.TryPop(out Token token) || token.Kind != kind)
```
I think that's very obvious. EatSingleCharacterToken expects a token kind, so logically this method peeks if the next token matches the requested kind. The RHS would not pass when asking for "," while the next token is ":".

maurei replied:
OK, it wasn't clear for me that these clauses catch bad client-side input. Now this makes sense

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/FilterParser.cs:
```diff
@@ -0,0 +1,321 @@
+
+            EatSingleCharacterToken(TokenKind.CloseParen);
+
+            if (leftTerm is ResourceFieldChainExpression leftChain)
```
It could be a count() function, as documented in filtering.md.

bart-degreed commented on src/JsonApiDotNetCore/Internal/QueryStrings/FilterQueryStringParameterReader.cs:
```diff
@@ -0,0 +1,154 @@
+        private readonly FilterParser _filterParser;
+
+        private readonly List<FilterExpression> _filtersInGlobalScope = new List<FilterExpression>();
+        private readonly Dictionary<ResourceFieldChainExpression, List<FilterExpression>> _filtersPerScope = new Dictionary<ResourceFieldChainExpression, List<FilterExpression>>();
```
Sharp! But that is not true. Because `?filter=X&filter=Y` is equivalent to `?filter=or(X,Y)` and this code ensures that works.

bart-degreed replied:
I'll add a comment just to make sure.

maurei commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/ResourceFieldChainResolver.cs:
```diff
@@ -0,0 +1,260 @@
+            {
+                var relationship = GetRelationship(publicName, resourceContext, path);
+
+                validateCallback?.Invoke(relationship, resourceContext, path);
```
In practise these validation callbacks seem to result in a coupling between the *client-side interaction with/violation of query parameter capabilities* and the *chain resolving functionality*. I feel there are two separate concerns here: (1) resolving resource field chains based on the apps initial conditions (models, exposed attributes and defined relationships), versus (2) any (disallowed) client-side interaction with those initial conditions through validation of attribute capabilities. I think it might be a good idea to separate those:
- `ResourceFieldChainResolver` looks like very suitable as a reusable component in other places where such resolving is also needed (eg client side functionality for future server-to-server features). When reused these validation callbacks probably won't have any meaning.
- These validation callbacks seem to string together `IQueryStringParameterReader` services, their embedded parsers  (`IncludeParser`, `FilterParser` etc) and finally the ResourceFieldChainResolver as these callbacks are passed from the former all the way down to the latter. 

We could consider pulling out the validation from the chain resolver and bring that to the `IQueryStringParameterReader` layer. Right now I'm not sure there is a very clear different responsibility for eg `IncludeQueryStringParameterReader` versus the `IncludeParser` and I think bringing the validation call to the former could lead to a better justification of these two being separate classes. I feel that the chain resolver shouldn't have to worry about client-side related validation anyway. The `IncludeQueryStringParameterReader` could just loop through the return value of `IncludeParser`. I don't think the performance implications of this are of any significance. ( [I'm doing something similar here
](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/e0270bd3cd344ebb5ff99593cc2c96e54dd17218/src/JsonApiDotNetCore/Serialization/Server/ResponseSerializer.cs#L121), could have assigned these links in an earlier nested loop, but decided to factor them out for separation purposes).

Additional effect of refactoring it like this would be that the constructor of the `QueryParser` descendants will be simplified. I think this is good because
- We can just do DI, which makes sense because `QueryParser` already has a a dependency that is DI-registered. 
- It will make the responsibility of `QueryParser` less ambiguous: right now the instantiation of QueryParser is done with a `validation callback` even though the instance is a `Parser`, not a `Validator`, and this constructor signature parameter ends up not even being called by the Parser itself but is passed along to a method call of a slightly ambiguous composite dependency (chain field resolver: ambiguous because it is manually instantiated in the abstract base class rather than injected).

bart-degreed replied:
I think you're mixing things up, which is understandable because I changed that recently. Initially, readers were responsible for dealing with the complexity of resolving chains. I moved that into the parsers (as an implmentation detail) to make the parsers easily reusable from resource definitions. `ResourceFieldChainResolver` and `FieldChainRequirements` are internal, non-shared implementation details because I'm likely going to change their interaction in the future. Or remove them entirely, I'm just not sure right now what would be the best way to expose their functionality. It would only be needed for the most-advanced scenarios like adding support for additional expressions, ie an Average function to be used in filters and sorting.

What's left in the readers now is only a callback that validates **a single field**. This separation is needed because resource capabilities should be taken into account when reading external data (query strings), but developers may conditionally override such limitations from resource definitions and have full freedom in doing so.

maurei commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/QueryParser.cs:
```diff
@@ -0,0 +1,91 @@
+
+        protected QueryParser(IResourceContextProvider resourceContextProvider)
+        {
+            ChainResolver = new ResourceFieldChainResolver(resourceContextProvider);
```
We can use DI here, making the dependencies of QueryParser less ambiguous and making `ResourceFieldChainResolver` reusable. See other review comment in which this is also mentioned

bart-degreed replied:
`ResourceFieldChainResolver` is not a candidate for reuse right now. That is why I pushed it into the parsers as an implementation detail that is subject to change.

bart-degreed commented on src/JsonApiDotNetCore/Queries/IQueryConstraintProvider.cs:
```diff
@@ -0,0 +1,15 @@
+    /// <summary>
+    /// Provides constraints (such as filters, sorting, pagination, sparse fieldsets and inclusions) to be applied on a data set.
+    /// </summary>
+    public interface IQueryConstraintProvider
```
My naming stems from the perspective of a data source engine like a relational database. Constraints limit the engine from just returning the table data as-is (which you get from `select * from table`). Constraints represent additional clauses that requires the engine to perform extra work, such as filtering, ordering, paging and projecting. Though debatable, including related data can also be considered extra work.

From a user perspective, I think the primary reason to implement a constraint provider is to block certain built-in behaviors.

I'm open to better names, but this was the best I could come up with.

maurei replied:
> Constraints represent additional clauses that requires the engine to perform extra work, such as filtering, ordering, paging and projecting

Perhaps *operation* is a good word to reflect the extra work performed by the engine. (`IQueryOperationProvider`, `IQueryParameterOperationProvider`?)

bart-degreed replied:
I like that, but it clashes with *atomic operations* which means something else entirely, making the terminology confusing.

maurei replied:
Thats a bummer indeed. Perhaps `IQueryParameterActions` , or we perhaps we could choose to go with having *atomic operations* and *query parameter operations* coexisting. (query expressions and linq expressions are also coexisting, I think it might not be too confusing)

bart-degreed replied:
Thanks for the suggestions. Action triggers the thought of a procedure with me, while these are declarative form. They express intent (what you want) instead of action (how to achieve that). That's the point of separation of intent and implementation being introduced here. And they are not parameters, but typically expand into clauses (which is another implementation detail). QueryExpression was an unavoidable naming clash, which I'd prefer to avoid whenever possible.

Given my explanation where the name came from I think it is not incorrect so I still think it is the best option, but maybe adding that as a comment can take away any confusion.

bart-degreed replied:
DataStoreConstraintProvider?

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/QueryParser.cs:
```diff
@@ -0,0 +1,91 @@
+    /// </summary>
+    public abstract class QueryParser
+    {
+        private protected ResourceFieldChainResolver ChainResolver { get; }
```
Because `ChainResolver`, at this time, is not intended to be used outside of JADNC internally. I'm not happy with how it works and want to be able to provide a better way in the future.

bart-degreed replied:
I've pushed a commit that explains the intent of `FieldChainRequirements`, hope that helps.

bart-degreed commented on src/JsonApiDotNetCore/Internal/Queries/Parsing/QueryParser.cs:
```diff
@@ -0,0 +1,91 @@
+    {
+        private protected ResourceFieldChainResolver ChainResolver { get; }
+
+        protected Stack<Token> TokenStack { get; private set; }
```
Your example points to a *field*, but this is a property. Protected properties should have Pascal casing as described in the Microsoft naming guidelines. I'm fine with changing the case for protected fields to comply with existing code here (please identity them if any), though I would more likely change that into a property as protected fields are discouraged in the Microsoft framework guidelines.

maurei commented on src/JsonApiDotNetCore/Serialization/Server/FieldsToSerialize.cs:
```diff
@@ -1,44 +1,74 @@
-        public List<AttrAttribute> GetAllowedAttributes(Type type, RelationshipAttribute relationship = null)
-        {   // get the list of all exposed attributes for the given type.
-            var allowed = _resourceGraph.GetAttributes(type);
+        public IReadOnlyCollection<AttrAttribute> GetAttributes(Type resourceType, RelationshipAttribute relationship = null)
```
Whereas before the to-be-rendered attrs was directly retrieved from the resource graph, now the attributes from a sparse fieldset selection is also weighed in. If I'm understanding correctly, this is required because else it would be possible that a field is included in the API output which *wasn't* selected in the sparse field selection, which would break the spec (or at least a recommendation?)

bart-degreed replied:
> Whereas before the to-be-rendered attrs were just retrieved from the resource graph, now the selected attributes from a sparse fieldset selection are also weighed in. 

That is incorrect, the old code also took sparse fieldsets into account (line 39).

bart-degreed replied:
See also explaining docs at `ResourceDefinition.OnApplySparseFieldSet`
